File: lib/pdp10-extint.c
/*
 * pdp10-extint.c
 * Copyright (C) 2013-2015  Mikael Pettersson
 *
 * This file is part of pdp10-tools.
 *
 * pdp10-tools is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * pdp10-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with pdp10-tools.  If not, see <http://www.gnu.org/licenses/>.
 */
/*
 * Provide types and procedures for converting 18 and 36-bit integers
 * to and from arrays of 9-bit bytes (nonets).  Use these together with
 * pdp10_fread() and pdp10_fwrite() to convert 18 and 36-bit integers
 * between host-level and file-level binary representations.
 */
#include "pdp10-extint.h"

/*
 * The behaviour of the PDP10's byte pointers implies a big-endian storage model,
 * as does the layout of its 72-bit long integers.
 */

void pdp10_uint18_to_ext(pdp10_uint18_t val, struct pdp10_ext_uint18 *ext)
{
    ext->nonet[0] = (val >> 9) & 0x1FF;
    ext->nonet[1] = val & 0x1FF;
}

pdp10_uint18_t pdp10_uint18_from_ext(const struct pdp10_ext_uint18 *ext)
{
    return
	((pdp10_uint18_t)(ext->nonet[0] & 0x1FF) << 9)
	| (ext->nonet[1] & 0x1FF);
}

void pdp10_uint36_to_ext(pdp10_uint36_t val, struct pdp10_ext_uint36 *ext)
{
    ext->nonet[0] = (val >> 27) & 0x1FF;
    ext->nonet[1] = (val >> 18) & 0x1FF;
    ext->nonet[2] = (val >> 9) & 0x1FF;
    ext->nonet[3] = val & 0x1FF;
}

pdp10_uint36_t pdp10_uint36_from_ext(const struct pdp10_ext_uint36 *ext)
{
    return
	((pdp10_uint36_t)(ext->nonet[0] & 0x1FF) << 27)
	| ((pdp10_uint36_t)(ext->nonet[1] & 0x1FF) << 18)
	| ((pdp10_uint36_t)(ext->nonet[2] & 0x1FF) << 9)
	| (ext->nonet[3] & 0x1FF);
}

File: erlang/apps/lib/src/pdp10_extint.erl
%%% -*- erlang-indent-level: 2 -*-
%%%
%%% converts between 18 and 36-bit integers and sequences of nonets
%%% Copyright (C) 2013-2025  Mikael Pettersson
%%%
%%% This file is part of pdp10-tools.
%%%
%%% pdp10-tools is free software: you can redistribute it and/or modify
%%% it under the terms of the GNU General Public License as published by
%%% the Free Software Foundation, either version 3 of the License, or
%%% (at your option) any later version.
%%%
%%% pdp10-tools is distributed in the hope that it will be useful,
%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%%% GNU General Public License for more details.
%%%
%%% You should have received a copy of the GNU General Public License
%%% along with pdp10-tools.  If not, see <http://www.gnu.org/licenses/>.

-module(pdp10_extint).

-export([ uint18_to_ext/1
        , uint18_from_ext/1
        , uint36_to_ext/1
        , uint36_from_ext/1
        ]).

-include_lib("lib/include/stdint.hrl").

%% The PDP10 is big-endian, so the conversions here are big-endian.
%% The bytes (nonets) are numbered in storage order.

-spec uint18_to_ext(uint18_t()) -> [uint9_t()].
uint18_to_ext(U18) ->
  [_B0 = (U18 bsr 9) band ?UINT9_MAX, _B1 = U18 band ?UINT9_MAX].

-spec uint18_from_ext([uint9_t()]) -> uint18_t().
uint18_from_ext([B0, B1]) ->
  ((B0 band ?UINT9_MAX) bsl 9) bor (B1 band ?UINT9_MAX).

-spec uint36_to_ext(uint36_t()) -> [uint9_t()].
uint36_to_ext(U36) ->
  [_B0 = (U36 bsr 27) band ?UINT9_MAX,
   _B1 = (U36 bsr 18) band ?UINT9_MAX,
   _B2 = (U36 bsr  9) band ?UINT9_MAX,
   _B3 =  U36         band ?UINT9_MAX].

-spec uint36_from_ext([uint9_t()]) -> uint36_t().
uint36_from_ext([B0, B1, B2, B3]) ->
  ((B0 band ?UINT9_MAX) bsl 27) bor
  ((B1 band ?UINT9_MAX) bsl 18) bor
  ((B2 band ?UINT9_MAX) bsl  9) bor
   (B3 band ?UINT9_MAX).
